### PR TITLE
Feature/database/duplication

### PR DESCRIPTION
**Pull request by sledzgrzegorz**
This PR contains changes associated with caching books and authors. When parser application starts working it performs some additional operations to avoid sending duplicated request to database for saving some objects.

### CODE REVIEW

File: database/src/main/java/repositories/GenreRepository.java [hunk 1]
```diff
@@ -1,21 +1,15 @@
...
     List<Genre> findAll();
 
-    default Genre findByName(String name) {
-        Optional<Genre> result = findAll().stream()
-                .filter(p -> p.getName().equals(name))
-                .findFirst();
-
-        return result.isPresent() ? result.get() : null;
-    }
+    Genre findByName(String name);
```
Comment (caroo7) [inline database/src/main/java/repositories/GenreRepository.java:13]: :+1: for repair my mistakes ;)

File: parser/src/main/java/main/ParserMain.java [hunk 2]
```diff
@@ -17,20 +16,19 @@
 
 import java.util.Set;
 
+@Log4j2
 public class ParserMain {
```
Comment (LazyScoundrel) [inline parser/src/main/java/main/ParserMain.java:20]: :+1:

File: parser/src/main/java/converter/BookDetailsToBookAssembler.java [hunk 3]
```diff
@@ -36,10 +36,18 @@
...
+
+    private List<Author> authorsCache;
+    private List<Book> booksCache;
+
+    public BookDetailsToBookAssembler(Mapper genreMapper, GenreRepository genreRepo, AuthorRepository authorRepo,BookRepository bookRepo) {
         this.genreMapper = genreMapper;
         this.genreRepo = genreRepo;
+        this.bookRepo=bookRepo;
         this.authorRepo = authorRepo;
+        this.authorsCache=authorRepo.findAll();
```
Comment (caroo7) [inline parser/src/main/java/converter/BookDetailsToBookAssembler.java]: move both cache creation to declaration place.

**Comment by LazyScoundrel** [comment]
:+1:

File: database/src/main/java/entities/Author.java [hunk 4]
```diff
@@ -23,4 +23,20 @@ public Author(String name) {
         this.name = name;
     }
 
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
```
Comment (egolesor) [inline database/src/main/java/entities/Author.java]: Performance issue! dont do that

File: parser/src/main/java/converter/BookDetailsToBookAssembler.java [hunk 5]
```diff
@@ -58,18 +66,27 @@ private Book convert(BookDetails bookDetails) {
         Set<Author> authors = retrieveAuthors(bookDetails);
         Set<Genre> genres = retrieveGenres(bookDetails);
 
-        return Book.builder().title(title).authors(authors).description(description).
+        Book book=Book.builder().title(title).authors(authors).description(description).
                 discount(discount).price(price).timestamp(timestamp).genres(genres).build();
+        if(booksCache.contains(book)){
```
Comment (egolesor) [inline parser/src/main/java/converter/BookDetailsToBookAssembler.java]: Do you test the hashCode ?and equals ?

File: parser/src/main/java/converter/BookDetailsToBookAssembler.java [hunk 6]
```diff
@@ -58,18 +66,27 @@ private Book convert(BookDetails bookDetails) {
...
     }
 
     private Set<Author> retrieveAuthors(BookDetails bookDetails) {
         Set<String> authorsString = bookDetails.getAuthors();
 
         return authorsString.stream().map(s -> {
-            Author author=authorRepo.findAuthor(s);
-            if (author == null) {
-                author = new Author(s);
+            Author author=new Author(s);
```
Comment (egolesor) [inline parser/src/main/java/converter/BookDetailsToBookAssembler.java]: Long lambda is not a good practice :snowflake:

File: parser/src/main/java/main/ParserMain.java [hunk 7]
```diff
@@ -17,20 +16,19 @@
...
-    private static Logger logger = LogManager.getRootLogger();
-
     public static void main(String[] args) {
-        logger.info("Parser application starts");
+        log.info("Parser application starts");
 
         if(args.length == 0) {
-            logger.error("No library name provided!");
+            log.error("No library name provided!");
             System.exit(-1);
```
Comment (egolesor) [inline parser/src/main/java/main/ParserMain.java:27]: System exit  ? :octopus: is there any reason to close the JVM in this situation ?

Comment (sledzgrzegorz) [reply parser/src/main/java/main/ParserMain.java:27]: It is intentionally usage

**Pull request merged**